man/cherryd.1
.\" Man page generated from reStructuredText.
.TH cherryd 1 "2009-06-15" "3.2.0" "web"
.SH NAME
cherryd \- Starts the CherryPy HTTP server as a daemon

.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level magin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..

.SH SYNOPSIS
\fBcherryd\fP [\-d] [\-f | \-s] [\-e ENV_NAME] [\-p PIDFILE_PATH] [\-P DIRPATH] [\-c CONFIG_FILE] \-i MODULE_NAME


.SH DESCRIPTION
\fBcherryd\fP is a Python script which starts the CherryPy webserver as a daemon.


.SH OPTIONS
.INDENT 0.0

.TP
.BI \-c\  CONFIG_FILE ,\ \-\-config\fn= CONFIG_FILE
Specifies a config file which is to be read and merged into the
CherryPy site\-wide config dict.  This option may be specified
multiple times.  For each CONFIG_FILE specified, \fBcherryd\fP will perform
\fBcherrypy.config.update()\fP.


.TP
.B \-d
Run the server as a daemon.


.TP
.BI \-e\  ENV_NAME ,\ \-\-environment\fn= ENV_NAME
Specifies the name of an environment to be applied.  An environment is a
canned set of configuration entries.  See \fI\%ENVIRONMENTS\fP below for a
list of the built\-in environments.


.TP
.B \-f
Start a fastcgi server instead of the default HTTP server.


.TP
.B \-s
Start a scgi server instead of the default HTTP server.


.TP
.BI \-i\  MODULE_NAME ,\ \-\-import\fn= MODULE_NAME
Specifies a module to import.  This option may be specified multiple times.
For each MODULE_NAME specified, \fBcherryd\fP will import the module.  This
is how you tell \fBcherryd\fP to run your application\'s startup code.
For all practical purposes, \fB\-i\fP is not optional; you will always need to
specify at least one module.


.TP
.BI \-p\  PIDFILE_PATH ,\ \-\-pidfile\fn= PIDFILE_PATH
Store the process id in PIDFILE_PATH.


.TP
.BI \-P\  DIRPATH ,\ \-\-Path\  DIRPATH
Specifies a directory to be inserted at the head of \fBsys.path\fP.  DIRPATH
should be an absolute path.  This option may be specified multiple times.
\fBcherryd\fP inserts all the specified DIRPATHs into \fBsys.path\fP before it
attempts to import modules specified with \fB\-i\fP.

.UNINDENT
For a terse summary of the options, run \fBcherryd \-\-help\fP.


.SH EXAMPLES
A site\-wide configuration file \fBsite.conf\fP:

.INDENT 0.0
.INDENT 3.5

[global]
.br
server.socket_host = "0.0.0.0"
.br
server.socket_port = 8008
.br
engine.autoreload.on = False
.br

.UNINDENT
.UNINDENT
The application startup code in \fBstartup.py\fP:

.INDENT 0.0
.INDENT 3.5

import cherrypy
.br
import my_controller
.br
cherrypy.log.error_file = \'/var/tmp/myapp\-error.log\'
.br
cherrypy.log.access_file = \'/var/tmp/myapp\-access.log\'
.br
config_root = { \'tools.encode.encoding\' : \'utf\-8\', }
.br
app_conf = { \'/\' : config_root }
.br
cherrypy.tree.mount(my_controller.Root(), script_name=\'\', config=app_conf)
.br

.UNINDENT
.UNINDENT
A corresponding \fBcherryd\fP command line:

.INDENT 0.0
.INDENT 3.5

cherryd \-d \-c site.conf \-i startup \-p /var/log/cherrypy/my_app.pid
.br

.UNINDENT
.UNINDENT

.SH DROPPING PRIVILEGES
If you want to serve your web application on TCP port 80 (or any port lower than
1024), the CherryPy HTTP server needs to start as root in order to bind to the
port.  Running a web application as root is reckless, so the application should
drop privileges from root to some other user and group.  The application must do
this itself, as \fBcherryd\fP does not do it for you.

To drop privileges, put the following lines into your startup code,
substituting appropriate values for \fBumask\fP, \fBuid\fP and \fBgid\fP:

.INDENT 0.0
.INDENT 3.5

from cherrypy.process.plugins import DropPrivileges
.br
DropPrivileges(cherrypy.engine, umask=022, uid=\'nobody\', gid=\'nogroup\').subscribe()
.br

.UNINDENT
.UNINDENT
Note that the values for \fBuid\fP and \fBgid\fP may be either user and group
names, or uid and gid integers.

Note that you must disable the engine Autoreload plugin, because the way
Autoreload works is by \fBexec()\fPing a new instance of the running process,
replacing the current instance.  Since root privileges are already dropped, the
new process instance will fail when it tries to perform a privileged operation
such as binding to a low\-numbered TCP port.


.SH ENVIRONMENTS
These are the built\-in environment configurations:


.SS staging
.INDENT 0.0
.INDENT 3.5

\'engine.autoreload.on\': False,
.br
\'checker.on\': False,
.br
\'tools.log_headers.on\': False,
.br
\'request.show_tracebacks\': False,
.br
\'request.show_mismatched_params\': False,
.br

.UNINDENT
.UNINDENT

.SS production
.INDENT 0.0
.INDENT 3.5

\'engine.autoreload_on\': False,
.br
\'checker.on\': False,
.br
\'tools.log_headers.on\': False,
.br
\'request.show_tracebacks\': False,
.br
\'request.show_mismatched_params\': False,
.br
\'log.screen\': False,
.br

.UNINDENT
.UNINDENT

.SS embedded
.INDENT 0.0
.INDENT 3.5

# For use with CherryPy embedded in another deployment stack, e.g. Apache mod_wsgi.
.br
\'engine.autoreload_on\': False,
.br
\'checker.on\': False,
.br
\'tools.log_headers.on\': False,
.br
\'request.show_tracebacks\': False,
.br
\'request.show_mismatched_params\': False,
.br
\'log.screen\': False,
.br
\'engine.SIGHUP\': None,
.br
\'engine.SIGTERM\': None,
.br

.UNINDENT
.UNINDENT

.SH BUGS
\fBcherryd\fP should probably accept command\-line options \fB\-\-uid\fP, \fB\-\-gid\fP, and
\fB\-\-umask\fP, and handle dropping privileges itself.


.SH AUTHOR
fumanchu

.nf
cherrypy.dev
.fi

.SH COPYRIGHT
This man page is placed in the public domain

.\" Generated by docutils manpage writer on 2009-06-15 15:07.
.\"


CHANGES.rst
v18.7.0
-------

* :pr:`1923`: Drop support for Python 3.5.

v18.6.1
-------

* :issue:`1849` via :pr:`1879`: Fixed XLF flag in gzip header
  emitted by gzip compression tool per
  :rfc:`1952#section-2.3.1` -- by :user:`webknjaz`.

* :issue:`1874`: Restricted depending on pywin32 only under
  CPython so that it won't get pulled-in under PyPy
  -- by :user:`webknjaz`.

* :issue:`1920`: Bumped minimum version of PyWin32 to 227.
  Block pywin32 install on Python 3.10 and later.

v18.6.0
-------

* :issue:`1776` via :pr:`1851`: Add support for UTF-8 encoded attachment
  file names in ``Content-Disposition`` header via :rfc:`6266#appendix-D`.

v18.5.0
-------

* :issue:`1827`: Fixed issue where bytes values in a ``HeaderMap``
  would be converted to strings.

* :pr:`1826`: Rely on
  `jaraco.collections <https://pypi.org/project/jaraco.collections>`_
  for its case-insensitive dictionary support.

v18.4.0
-------

* :pr:`1715`: Fixed issue in cpstats where the ``data/`` endpoint
  would fail with encoding errors on Python 3.

* :pr:`1821`: Simplify the passthrough of parameters to
  ``CPWebCase.getPage`` to cheroot. CherryPy now requires
  cheroot 8.2.1 or later.

v18.3.0
-------

* :pr:`1806`: Support handling multiple exceptions when processing hooks as
  reported in :issue:`1770`.

v18.2.0
-------

* File-based sessions no longer attempt to remove the lock files
  when releasing locks, instead deferring to the default behavior
  of zc.lockfile. Fixes :issue:`1391` and :issue:`1779`.

* :pr:`1794`: Add native support for ``308 Permanent Redirect``
  usable via ``raise cherrypy.HTTPRedirect('/new_uri', 308)``.

v18.1.2
-------

* Fixed :issue:`1377` via :pr:`1785`: Restore a native WSGI-less
  HTTP server support.
* :pr:`1769`: Reduce log level for non-error events in win32.py

v18.1.1
-------

* :pr:`1774` reverts :pr:`1759` as new evidence emerged that
  the original behavior was intentional. Re-opens :issue:`1758`.

v18.1.0
-------

* :issue:`1758` via :pr:`1759`: In the bus, when awaiting a
  state change, only publish after the state has changed.

v18.0.1
-------

* :issue:`1738` via :pr:`1736`: Restore support for 'bytes'
  in response headers.

* Substantial removal of Python 2 compatibility code.

v18.0.0
-------

* :issue:`1730`: Drop support for Python 2.7. CherryPy 17 will
  remain an LTS release for bug and security fixes.

* Drop support for Python 3.4.

v17.4.2
-------

* Fixed :issue:`1377` by backporting :pr:`1785` via :pr:`1786`:
  Restore a native WSGI-less HTTP server support.

v17.4.1
-------

* :issue:`1738` via :pr:`1755`: Restore support for 'bytes'
  in response headers (backport from v18.0.1).

v17.4.0
-------

* :commit:`a95e619f`: When setting Response Body, reject Unicode
  values, making behavior on Python 2 same as on Python 3.

* Other inconsequential refactorings.

v17.3.0
-------

* :issue:`1193` via :pr:`1729`: Rely on zc.lockfile for
  session concurrency support.

v17.2.0
-------

* :issue:`1690` via :pr:`1692`: Prevent orphaned Event object in cached
  304 response.

v17.1.0
-------

* :issue:`1694` via :pr:`1695`: Add support for accepting uploaded files
  with non-ascii filenames per RFC 5987.

v17.0.0
-------

* :issue:`1673`: CherryPy now allows namespace packages for
  its dependencies. Environments that cannot handle namespace
  packgaes like py2exe will need to add such support or pin to
  older CherryPy versions.

v16.0.3
-------

* :issue:`1722`: Pinned the ``tempora`` dependency against
  version 1.13 to avoid pulling in namespace packages.

v16.0.2
-------

* :issue:`1716` via :pr:`1717`: Fixed handling of url-encoded parameters
  in digest authentication handling, correcting regression in v14.2.0.

* :issue:`1719` via :commit:`1d41828`: Digest-auth tool will now return
  a status code of 401 for when a scheme other than 'digest' is
  indicated.

v16.0.0
-------

* :issue:`1688` via :commit:`38ad1da`: Removed  ``basic_auth`` and
  ``digest_auth`` tools and the ``httpauth`` module, which have been
  officially deprecated earlier in v14.0.0.

* Removed deprecated properties:

  - ``cherrypy._cpreqbody.Entity.type`` deprecated in favor of
    :py:attr:`cherrypy._cpreqbody.Entity.content_type`

  - ``cherrypy._cprequest.Request.body_params`` deprecated in favor of
    :py:attr:`cherrypy._cprequest.RequestBody.params`

* :issue:`1377`: In _cp_native server, set ``req.status`` using bytes
  (fixed in :pr:`1712`).

* :issue:`1697` via :commit:`841f795`: Fixed error on Python 3.7 with
  AutoReloader when ``__file__`` is ``None``.

* :issue:`1713` via :commit:`15aa80d`: Fix warning emitted during
  test run.

* :issue:`1370` via :commit:`38f199c`: Fail with HTTP 400 for invalid
  headers.

v15.0.0
-------

* :issue:`1708`: Removed components from webtest that were
  removed in the refactoring of cheroot.test.webtest for
  cheroot 6.1.0.

v14.2.0
-------

* :issue:`1680` via :pr:`1683`: Basic Auth and Digest Auth
  tools now support :rfc:`7617` UTF-8 charset decoding where
  possible, using latin-1 as a fallback.

v14.1.0
-------

* :cr-pr:`37`: Add support for peercreds lookup over UNIX domain socket.
  This enables app to automatically identify "who's on the other
  end of the wire".

  This is how you enable it::

    server.peercreds: True
    server.peercreds_resolve: True

  The first option will put remote numeric data to WSGI env vars:
  app's PID, user's id and group.

  Second option will resolve that into user and group names.

  To prevent expensive syscalls, data is cached on per connection
  basis.

v14.0.1
-------

* :issue:`1700`: Improve windows pywin32 dependency declaration via
  conditional extras.

v14.0.0
-------

* :issue:`1688`: Officially deprecated ``basic_auth`` and ``digest_auth``
  tools and the ``httpauth`` module, triggering DeprecationWarnings
  if they're used. Applications should instead adapt to use the
  more recent ``auth_basic`` and ``auth_digest`` tools.

cherrypy/cherryd.py
#! /usr/bin/env python
"""The CherryPy daemon."""

import getopt
import sys

import cherrypy
from cherrypy.process import plugins


shortopts = ["d", "e"]
longopts = []

# Help for restd command-line options.
help = """
cherryd. Start the cherrypy daemon.

Usage:
    cherryd <config filename>

"""


def start(options, configfile, *args):
    engine = cherrypy.engine
    
    siteconf = {}
    cherrypy._cpconfig.merge(siteconf, configfile)
    
    cherrypy.config.update(configfile)
    
    if options.has_key('-e'):
        cherrypy.config.update({'environment': options['-e']})
    
    # TODO: Make sure that log files are configurable from the conf file.
    
    # Only daemonize if asked to.
    if options.has_key('-d'):
        # Don't print anything to stdout/sterr.
        cherrypy.config.update({'log.screen': False})
        plugins.Daemonizer(engine).subscribe()
    
    if options.has_key('--pidfile'):
        plugins.PIDFile(engine, options['--pidfile']).subscribe()
    
    cherrypy.signal_handler.subscribe()
    
    # TODO: call a 'site setup' function (probably passing it siteconf).
    
    if options.has_key('--fastcgi'):
        # turn off autoreload when using fastcgi
        cherrypy.config.update({'autoreload.on': False})
        
        cherrypy.server.unsubscribe()
        
        fastcgi_port = cherrypy.config.get('server.socket_port', 4000)
        fastcgi_bindaddr = cherrypy.config.get('server.socket_host', '0.0.0.0')
        bindAddress = (fastcgi_bindaddr, fastcgi_port)
        try:
            # Always start the engine; this will start all other services
            engine.start()
            
            from flup.server.fcgi import WSGIServer
            engine.log('Serving FastCGI on %s:%d' % bindAddress)
            engine.fcgi = WSGIServer(application=wsgiapp,
                                     bindAddress=bindAddress)
            engine.fcgi.run()
            engine.log('FastCGI Server on %s:%d shut down' % bindAddress)
        finally:
            engine.stop()
    else:
        # Always start the engine; this will start all other services
        s = cherrypy.server
        s.httpserver, s.bind_addr = s.httpserver_from_self()
        s.httpserver.wsgi_app = wsgiapp
        engine.start()
        engine.block()


if __name__ == '__main__':
    try:
        opts, args = getopt.getopt(sys.argv[1:], shortopts, longopts)
    except getopt.GetoptError:
        print help
        sys.exit(2)
    
    start(dict(opts), *tuple(args))
